### PR TITLE
Aztec Code: decode ECI and FNC1

### PR DESCRIPTION
**Pull request by dlenski**
Addresses #1327.

- Decode FNC1 (as `0x1d`/escape character; same as Code128, DataMatrix, others appear to do)
- Decode ECI (modeled on PDF417 `DecodedBitStreamParser`)
- Fix Aztec Code encoder tests
- Include `ecLevel` estimation (percentage) in Aztec `DecodedResult`
- Add tests which specifically cover Aztec ECI decoding?

TODO:

- Handle Aztec ECI **en**coding?

### CODE REVIEW

File: core/src/test/java/com/google/zxing/aztec/encoder/EncoderTest.java [hunk 1]
```diff
@@ -577,15 +577,21 @@ private static void testHighLevelEncodeString(String s, String expectedBits) {
     BitArray bits = new HighLevelEncoder(s.getBytes(StandardCharsets.ISO_8859_1)).encode();
     String receivedBits = stripSpace(bits.toString());
     assertEquals("highLevelEncode() failed for input string: " + s, stripSpace(expectedBits), receivedBits);
-    assertEquals(s, Decoder.highLevelDecode(toBooleanArray(bits)));
+    try {
+      assertEquals(s, Decoder.highLevelDecode(toBooleanArray(bits)));
+    } catch (FormatException e) {
```
Comment (dlenski) [inline core/src/test/java/com/google/zxing/aztec/encoder/EncoderTest.java:582]: Not sure if there's another way that you'd prefer this to be handled @srowen.

Comment (srowen) [reply core/src/test/java/com/google/zxing/aztec/encoder/EncoderTest.java:582]: Just let the method throw FormatException. We want it to fail the test if unexpected.

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java [hunk 2]
```diff
@@ -92,11 +96,15 @@ public static String highLevelDecode(boolean[] correctedBits) {
...
-  private static String getEncodedData(boolean[] correctedBits) {
+  private static String getEncodedData(boolean[] correctedBits) throws FormatException {
     int endIndex = correctedBits.length;
     Table latchTable = Table.UPPER; // table most recently latched to
     Table shiftTable = Table.UPPER; // table to use for the next read
-    StringBuilder result = new StringBuilder(20);
+
+    ByteArrayOutputStream decodedBytes = new ByteArrayOutputStream(); // buffer of decoded bytes
+    Charset encoding = StandardCharsets.ISO_8859_1; // current character encoding (set by ECI)
+    StringBuilder result = new StringBuilder(correctedBits.length >> 2); // final string result
```
Comment (srowen) [inline core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:106]: Maybe just divide by 4 for clarity

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java [hunk 3]
```diff
@@ -142,12 +185,14 @@ private static String getEncodedData(boolean[] correctedBits) {
...
         } else {
-          result.append(str);
+          byte[] b = str.getBytes(StandardCharsets.US_ASCII);
+          decodedBytes.write(b, 0, b.length);
           // Go back to whatever mode we had been in
           shiftTable = latchTable;
         }
       }
     }
+    result.append(new String(decodedBytes.toByteArray(), encoding));
```
Comment (srowen) [inline core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:195]: Does this need to check whether the byte array is empty? if it results in an empty string, fine

**Review by srowen — CHANGES_REQUESTED** [review]
Encoding can be left to another change. We do probably need a test here, at least one.

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java [hunk 4]
```diff
@@ -196,13 +241,23 @@ private static String getCharacter(Table table, int code) {
     }
   }
 
+  static final class CorrectedBitsResult {
+    private final boolean[] correctBits;
+    private final String ecLevel;
```
[the remark this answers is not in the document]
Comment (dlenski) [reply core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:246]: That's correct. I can make it an `int` here, but the `DecodedResult` will still expect it as a `String`

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java (hunk 2 above)
Comment (dlenski) [reply core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:106]: :+1:

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java [hunk 5]
```diff
@@ -131,7 +139,42 @@ private static String getEncodedData(boolean[] correctedBits) {
...
+            break;
+          }
+          int n = readCode(correctedBits, index, 3);
+          index += 3;
+          switch (n) {
+            case 0:
+              result.append((char) 29);  // translate FNC1 as ASCII 29
+              break;
+            case 7:
+              throw new IllegalArgumentException("FLG(7) is reserved and illegal");
```
[the remark this answers is not in the document]
Comment (dlenski) [reply core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:153]: :+1:. I was seeing `IllegalArgumentException` used in the QR decoder, and wasn't sure which was best.

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java (hunk 3 above)
Comment (dlenski) [reply core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:195]: > Does this need to check whether the byte array is empty?

Yes, it works fine with an empty string.

File: core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (srowen) [reply core/src/main/java/com/google/zxing/aztec/decoder/Decoder.java:153]: Yeah I think I tended to use IAE for caller errors to APIs, but not for problems decoding barcodes. Not sure I got that 100% right but the decoders shouldn't generally throw IAE

**Comment by srowen** [comment]
Looks good, thank you.

**Pull request merged**